File: client/src/achievements/catalog.ts
import { parseCanonical } from "../numeric";

export const ACHIEVEMENT_CATALOG_SCHEMA_VERSION = 1 as const;
export const MAX_EXACT_INTEGER = 9_007_199_254_740_991 as const;

export type AchievementScope = "run" | "career";
export type AchievementCondition =
  | Readonly<{ kind: "fact_present"; factKind: string }>
  | Readonly<{ kind: "counter_at_least"; counter: string; minimum: number }>
  | Readonly<{ kind: "exit_count_at_least"; count: number }>
  | Readonly<{ kind: "owns_generator_at_least"; generatorId: string; count: number }>
  | Readonly<{ kind: "all_of"; conditions: readonly AchievementCondition[] }>;

export type AchievementProof =
  | Readonly<{ kind: "provenance"; eventKinds: readonly string[] }>
  | Readonly<{ kind: "burn"; eventKind: string; resourceId: string; minimum: string }>
  | Readonly<{ kind: "possession"; justificationCopyKey: string }>;

export interface AchievementDefinition {
  readonly id: string;
  readonly conditionScope: AchievementScope;
  readonly condition: AchievementCondition;
  readonly proof: AchievementProof;
  readonly scoreGrant: number;
  readonly copyKey: string;
}

export interface AchievementCatalog {
  readonly schemaVersion: 1;
  readonly definitions: readonly AchievementDefinition[];
  readonly byId: ReadonlyMap<string, AchievementDefinition>;
}

export interface AchievementRegistry {
  readonly copyKeys: ReadonlySet<string>;
  readonly generatorIds: ReadonlySet<string>;
  readonly eventKinds: ReadonlySet<string>;
  readonly resourceIds: ReadonlySet<string>;
  readonly runCounters: ReadonlySet<string>;
  readonly careerCounters: ReadonlySet<string>;
  readonly provenanceSources: ReadonlyMap<string, readonly string[]>;
}

const mechanicalId = /^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$/;
const maximumConditionDepth = 4;
const maximumConditionNodes = 64;

export function loadAchievementCatalog(bytes: string | Uint8Array, registry: AchievementRegistry): AchievementCatalog {
  validateRegistry(registry);
  const source = typeof bytes === "string" ? bytes : new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  let parsed: unknown;
  try { parsed = JSON.parse(source); } catch (error) { syntax(`achievement catalog is invalid JSON: ${String(error)}`); }
  const root = exactObject(parsed, ["schema_version", "achievements"], "achievement catalog");
  if (root.schema_version !== ACHIEVEMENT_CATALOG_SCHEMA_VERSION || !Array.isArray(root.achievements)) syntax("achievement catalog has invalid root");
  let lastId = "";
  const definitions = root.achievements.map((raw, index): AchievementDefinition => {
    const label = `achievement catalog.achievements[${index}]`;
    const row = exactObject(raw, ["id", "condition_scope", "condition", "proof", "score_grant", "copy_key"], label);
    const id = identifier(row.id, `${label}.id`);
    if (id <= lastId || row.condition_scope !== "run" && row.condition_scope !== "career" || !registry.copyKeys.has(identifier(row.copy_key, `${label}.copy_key`))) syntax(`${label} is invalid`);
    lastId = id;
    const scope = row.condition_scope;
    const parsedCondition = parseCondition(row.condition, scope, registry, 1);
    if (parsedCondition.nodes > maximumConditionNodes) syntax(`${label}.condition is too large`);
    return Object.freeze({ id, conditionScope: scope, condition: parsedCondition.condition, proof: parseProof(row.proof, scope, parsedCondition.condition, registry), scoreGrant: integer(row.score_grant, 1, MAX_EXACT_INTEGER, `${label}.score_grant`), copyKey: row.copy_key as string });
  });
  return Object.freeze({ schemaVersion: 1, definitions: Object.freeze(definitions), byId: new Map(definitions.map((value) => [value.id, value])) });
}

function parseCondition(value: unknown, scope: AchievementScope, registry: AchievementRegistry, depth: number): { condition: AchievementCondition; nodes: number } {
  if (depth > maximumConditionDepth || value === null || typeof value !== "object" || Array.isArray(value)) syntax("achievement condition is invalid");
  const kind = (value as Record<string, unknown>).kind;
  if (kind === "fact_present") {
    const row = exactObject(value, ["kind", "fact_kind"], "fact_present");
    return { condition: Object.freeze({ kind, factKind: identifier(row.fact_kind, "fact_present.fact_kind") }), nodes: 1 };
  }
  if (kind === "counter_at_least") {
    const row = exactObject(value, ["kind", "counter", "minimum"], "counter_at_least");
    const counter = identifier(row.counter, "counter_at_least.counter");
    if (!(scope === "run" ? registry.runCounters : registry.careerCounters).has(counter)) syntax("counter is not registered for its scope");
    return { condition: Object.freeze({ kind, counter, minimum: integer(row.minimum, 0, MAX_EXACT_INTEGER, "counter_at_least.minimum") }), nodes: 1 };
  }
  if (kind === "exit_count_at_least") {
    const row = exactObject(value, ["kind", "count"], "exit_count_at_least");
    if (scope !== "career") syntax("exit_count_at_least requires career scope");
    return { condition: Object.freeze({ kind, count: integer(row.count, 1, MAX_EXACT_INTEGER, "exit_count_at_least.count") }), nodes: 1 };
  }
  if (kind === "owns_generator_at_least") {
    const row = exactObject(value, ["kind", "generator_id", "count"], "owns_generator_at_least");
    const generatorId = identifier(row.generator_id, "owns_generator_at_least.generator_id");
    if (scope !== "run" || !registry.generatorIds.has(generatorId)) syntax("owns_generator_at_least requires a registered run generator");
    return { condition: Object.freeze({ kind, generatorId, count: integer(row.count, 1, MAX_EXACT_INTEGER, "owns_generator_at_least.count") }), nodes: 1 };
  }
  if (kind === "all_of") {
    const row = exactObject(value, ["kind", "conditions"], "all_of");
    if (!Array.isArray(row.conditions) || row.conditions.length < 2 || row.conditions.length > 16) syntax("all_of conditions are invalid");
    let nodes = 1;
    const conditions = row.conditions.map((child) => { const result = parseCondition(child, scope, registry, depth + 1); nodes += result.nodes; return result.condition; });
    return { condition: Object.freeze({ kind, conditions: Object.freeze(conditions) }), nodes };
  }
  return syntax("achievement condition kind is invalid");
}

function parseProof(value: unknown, scope: AchievementScope, condition: AchievementCondition, registry: AchievementRegistry): AchievementProof {
  if (value === null || typeof value !== "object" || Array.isArray(value)) syntax("achievement proof is invalid");
  const kind = (value as Record<string, unknown>).kind;
  if (kind === "provenance") {
    const row = exactObject(value, ["kind", "event_kinds"], "provenance proof");
    if (!Array.isArray(row.event_kinds) || row.event_kinds.length === 0 || containsKind(condition, "owns_generator_at_least")) syntax("provenance proof is incompatible");
    const eventKinds = row.event_kinds.map((item) => identifier(item, "provenance.event_kinds"));
    if (eventKinds.some((item, index) => !registry.eventKinds.has(item) || index > 0 && eventKinds[index - 1]! >= item)) syntax("provenance event kinds are invalid");
    const declared = new Set(eventKinds);
    for (const source of provenanceSourceKeys(condition, scope)) {
      const required = registry.provenanceSources.get(source);
      if (!required || required.length === 0 || required.some((eventKind) => !declared.has(eventKind))) syntax("provenance proof does not derive its condition");
    }
    return Object.freeze({ kind, eventKinds: Object.freeze(eventKinds) });
  }
  if (kind === "burn") {
    const row = exactObject(value, ["kind", "event_kind", "resource_id", "minimum"], "burn proof");
    const eventKind = identifier(row.event_kind, "burn.event_kind");
    const resourceId = identifier(row.resource_id, "burn.resource_id");
    if (scope !== "run" || !registry.eventKinds.has(eventKind) || !registry.resourceIds.has(resourceId) || typeof row.minimum !== "string") syntax("burn proof is incompatible");
    const minimum = parseCanonical(row.minimum);
    if (!minimum.gt(0)) syntax("burn minimum must be positive");
    return Object.freeze({ kind, eventKind, resourceId, minimum: minimum.toString() });
  }
  if (kind === "possession") {
    const row = exactObject(value, ["kind", "justification_copy_key"], "possession proof");
    const justificationCopyKey = identifier(row.justification_copy_key, "possession.justification_copy_key");
    if (scope !== "run" || !containsKind(condition, "owns_generator_at_least") || !registry.copyKeys.has(justificationCopyKey)) syntax("possession proof is incompatible");
    return Object.freeze({ kind, justificationCopyKey });
  }
  return syntax("achievement proof kind is invalid");
}

function containsKind(condition: AchievementCondition, kind: AchievementCondition["kind"]): boolean {
  return condition.kind === kind || condition.kind === "all_of" && condition.conditions.some((child) => containsKind(child, kind));
}

function provenanceSourceKeys(condition: AchievementCondition, scope: AchievementScope): readonly string[] {
  switch (condition.kind) {
    case "fact_present": return [`fact:${condition.factKind}`];
    case "counter_at_least": return [`counter:${scope}:${condition.counter}`];
    case "exit_count_at_least": return ["exit_count"];
    case "owns_generator_at_least": return [];
    case "all_of": return condition.conditions.flatMap((child) => provenanceSourceKeys(child, scope)).sort();
  }
}

function validateRegistry(registry: AchievementRegistry): void {
  for (const values of [registry.copyKeys, registry.generatorIds, registry.eventKinds, registry.resourceIds, registry.runCounters, registry.careerCounters]) {
    for (const value of values) identifier(value, "achievement registry key");
  }
  for (const [source, kinds] of registry.provenanceSources) {
    if (!validProvenanceSource(source) || kinds.length === 0 || kinds.some((kind, index) => !registry.eventKinds.has(kind) || !mechanicalId.test(kind) || index > 0 && kinds[index - 1]! >= kind)) syntax("achievement provenance source registry is invalid");
  }
}

function validProvenanceSource(value: string): boolean {
  if (value === "exit_count") return true;
  const parts = value.split(":");
  if (parts.length === 2 && parts[0] === "fact") return mechanicalId.test(parts[1]!);
  return parts.length === 3 && parts[0] === "counter" && (parts[1] === "run" || parts[1] === "career") && mechanicalId.test(parts[2]!);
}

function exactObject(value: unknown, keys: readonly string[], label: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) syntax(`${label} must be an object`);
  const actual = Object.keys(value).sort(); const expected = [...keys].sort();
  if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index])) syntax(`${label} has invalid keys`);
  return value as Record<string, unknown>;
}

function identifier(value: unknown, label: string): string {
  if (typeof value !== "string" || !mechanicalId.test(value)) syntax(`${label} is invalid`);
  return value;
}

function integer(value: unknown, minimum: number, maximum: number, label: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < minimum || (value as number) > maximum) syntax(`${label} is invalid`);
  return value as number;
}

function syntax(message: string): never { throw new SyntaxError(message); }

File: server/achievements/catalog.go
// Package achievements owns permanent achievement IDs and their exact score.
// Achievement score is neither an economy resource nor a Clout mint.
package achievements

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"regexp"
	"sort"

	"cloud-clicker/server/decimal"
)

const (
	CatalogSchemaVersion  = 1
	maximumConditionDepth = 4
	maximumConditionNodes = 64
)

var (
	ErrInvalidCatalog = errors.New("invalid achievement catalog")
	mechanicalID      = regexp.MustCompile(`^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$`)
)

type ConditionScope string

const (
	ScopeRun    ConditionScope = "run"
	ScopeCareer ConditionScope = "career"
)

type ConditionKind string

const (
	ConditionFactPresent          ConditionKind = "fact_present"
	ConditionCounterAtLeast       ConditionKind = "counter_at_least"
	ConditionExitCountAtLeast     ConditionKind = "exit_count_at_least"
	ConditionOwnsGeneratorAtLeast ConditionKind = "owns_generator_at_least"
	ConditionAllOf                ConditionKind = "all_of"
)

type ProofKind string

const (
	ProofProvenance ProofKind = "provenance"
	ProofBurn       ProofKind = "burn"
	ProofPossession ProofKind = "possession"
)

type Condition struct {
	Kind        ConditionKind
	FactKind    string
	Counter     string
	Minimum     int64
	GeneratorID string
	Children    []Condition
}

type Proof struct {
	Kind                 ProofKind
	EventKinds           []string
	EventKind            string
	ResourceID           string
	Minimum              string
	JustificationCopyKey string
}

type Definition struct {
	ID             string
	ConditionScope ConditionScope
	Condition      Condition
	Proof          Proof
	ScoreGrant     int64
	CopyKey        string
}

type Registry struct {
	CopyKeys       map[string]bool
	GeneratorIDs   map[string]bool
	EventKinds     map[string]bool
	ResourceIDs    map[string]bool
	RunCounters    map[string]bool
	CareerCounters map[string]bool
}

type Catalog struct {
	Definitions []Definition
	byID        map[string]Definition
}

type rawCatalog struct {
	SchemaVersion int             `json:"schema_version"`
	Achievements  []rawDefinition `json:"achievements"`
}

type rawDefinition struct {
	ID             string          `json:"id"`
	ConditionScope ConditionScope  `json:"condition_scope"`
	Condition      json.RawMessage `json:"condition"`
	Proof          json.RawMessage `json:"proof"`
	ScoreGrant     int64           `json:"score_grant"`
	CopyKey        string          `json:"copy_key"`
}

type rawCondition struct {
	Kind        ConditionKind     `json:"kind"`
	FactKind    string            `json:"fact_kind,omitempty"`
	Counter     string            `json:"counter,omitempty"`
	Minimum     *int64            `json:"minimum,omitempty"`
	Count       *int64            `json:"count,omitempty"`
	GeneratorID string            `json:"generator_id,omitempty"`
	Conditions  []json.RawMessage `json:"conditions,omitempty"`
}

type rawProof struct {
	Kind                 ProofKind `json:"kind"`
	EventKinds           []string  `json:"event_kinds,omitempty"`
	EventKind            string    `json:"event_kind,omitempty"`
	ResourceID           string    `json:"resource_id,omitempty"`
	Minimum              string    `json:"minimum,omitempty"`
	JustificationCopyKey string    `json:"justification_copy_key,omitempty"`
}

func LoadCatalog(data []byte, registry Registry) (*Catalog, error) {
	if !validRegistry(registry) {
		return nil, ErrInvalidCatalog
	}
	var raw rawCatalog
	if err := decodeStrict(data, &raw); err != nil || raw.SchemaVersion != CatalogSchemaVersion || raw.Achievements == nil {
		return nil, fmt.Errorf("%w: root", ErrInvalidCatalog)
	}
	catalog := &Catalog{Definitions: make([]Definition, 0, len(raw.Achievements)), byID: map[string]Definition{}}
	lastID := ""
	for index, source := range raw.Achievements {
		if !mechanicalID.MatchString(source.ID) || source.ID <= lastID || (source.ConditionScope != ScopeRun && source.ConditionScope != ScopeCareer) ||
			source.ScoreGrant < 1 || source.ScoreGrant > decimal.MaxExactInteger || !registry.CopyKeys[source.CopyKey] {
			return nil, fmt.Errorf("%w: achievements[%d]", ErrInvalidCatalog, index)
		}
		condition, nodes, err := parseCondition(source.Condition, source.ConditionScope, registry, 1)
		if err != nil || nodes > maximumConditionNodes {
			return nil, fmt.Errorf("%w: achievements[%d].condition", ErrInvalidCatalog, index)
		}
		proof, err := parseProof(source.Proof, source.ConditionScope, condition, registry)
		if err != nil {
			return nil, fmt.Errorf("%w: achievements[%d].proof", ErrInvalidCatalog, index)
		}
		definition := Definition{ID: source.ID, ConditionScope: source.ConditionScope, Condition: condition, Proof: proof, ScoreGrant: source.ScoreGrant, CopyKey: source.CopyKey}
		catalog.Definitions = append(catalog.Definitions, definition)
		catalog.byID[definition.ID], lastID = definition, definition.ID
	}
	return catalog, nil
}

func parseCondition(data []byte, scope ConditionScope, registry Registry, depth int) (Condition, int, error) {
	if depth > maximumConditionDepth {
		return Condition{}, 0, ErrInvalidCatalog
	}
	var discriminator struct {
		Kind ConditionKind `json:"kind"`
	}
	if json.Unmarshal(data, &discriminator) != nil {
		return Condition{}, 0, ErrInvalidCatalog
	}
	var raw rawCondition
	if err := decodeStrict(data, &raw); err != nil {
		return Condition{}, 0, err
	}
	switch raw.Kind {
	case ConditionFactPresent:
		if !mechanicalID.MatchString(raw.FactKind) || raw.Counter != "" || raw.Minimum != nil || raw.Count != nil || raw.GeneratorID != "" || raw.Conditions != nil {
			return Condition{}, 0, ErrInvalidCatalog
		}
		return Condition{Kind: raw.Kind, FactKind: raw.FactKind}, 1, nil
	case ConditionCounterAtLeast:
		counters := registry.RunCounters
		if scope == ScopeCareer {
			counters = registry.CareerCounters
		}
		if !counters[raw.Counter] || raw.Minimum == nil || *raw.Minimum < 0 || *raw.Minimum > decimal.MaxExactInteger || raw.FactKind != "" || raw.Count != nil || raw.GeneratorID != "" || raw.Conditions != nil {
			return Condition{}, 0, ErrInvalidCatalog
		}
		return Condition{Kind: raw.Kind, Counter: raw.Counter, Minimum: *raw.Minimum}, 1, nil
	case ConditionExitCountAtLeast:
		if scope != ScopeCareer || raw.Count == nil || *raw.Count < 1 || *raw.Count > decimal.MaxExactInteger || raw.FactKind != "" || raw.Counter != "" || raw.Minimum != nil || raw.GeneratorID != "" || raw.Conditions != nil {
			return Condition{}, 0, ErrInvalidCatalog
		}
		return Condition{Kind: raw.Kind, Minimum: *raw.Count}, 1, nil
	case ConditionOwnsGeneratorAtLeast:
		if scope != ScopeRun || !registry.GeneratorIDs[raw.GeneratorID] || raw.Count == nil || *raw.Count < 1 || *raw.Count > decimal.MaxExactInteger || raw.FactKind != "" || raw.Counter != "" || raw.Minimum != nil || raw.Conditions != nil {
			return Condition{}, 0, ErrInvalidCatalog
		}
		return Condition{Kind: raw.Kind, GeneratorID: raw.GeneratorID, Minimum: *raw.Count}, 1, nil
	case ConditionAllOf:
		if len(raw.Conditions) < 2 || len(raw.Conditions) > 16 || raw.FactKind != "" || raw.Counter != "" || raw.Minimum != nil || raw.Count != nil || raw.GeneratorID != "" {
			return Condition{}, 0, ErrInvalidCatalog
		}
		result := Condition{Kind: raw.Kind, Children: make([]Condition, len(raw.Conditions))}
		nodes := 1
		for index, child := range raw.Conditions {
			parsed, childNodes, err := parseCondition(child, scope, registry, depth+1)
			if err != nil {
				return Condition{}, 0, err
			}
			result.Children[index], nodes = parsed, nodes+childNodes
		}
		return result, nodes, nil
	default:
		return Condition{}, 0, ErrInvalidCatalog
	}
}

func parseProof(data []byte, scope ConditionScope, condition Condition, registry Registry) (Proof, error) {
	var raw rawProof
	if err := decodeStrict(data, &raw); err != nil {
		return Proof{}, err
	}
	switch raw.Kind {
	case ProofProvenance:
		if len(raw.EventKinds) == 0 || raw.EventKind != "" || raw.ResourceID != "" || raw.Minimum != "" || raw.JustificationCopyKey != "" || containsKind(condition, ConditionOwnsGeneratorAtLeast) {
			return Proof{}, ErrInvalidCatalog
		}
		last := ""
		for _, kind := range raw.EventKinds {
			if kind <= last || !registry.EventKinds[kind] {
				return Proof{}, ErrInvalidCatalog
			}
			last = kind
		}
		return Proof{Kind: raw.Kind, EventKinds: append([]string(nil), raw.EventKinds...)}, nil
	case ProofBurn:
		if scope != ScopeRun || !registry.EventKinds[raw.EventKind] || !registry.ResourceIDs[raw.ResourceID] || raw.EventKinds != nil || raw.JustificationCopyKey != "" {
			return Proof{}, ErrInvalidCatalog
		}
		minimum, err := decimal.ParseCanonical(raw.Minimum)
		if err != nil || !minimum.IsStateValue() || !minimum.Gt(decimal.Zero) {
			return Proof{}, ErrInvalidCatalog
		}
		return Proof{Kind: raw.Kind, EventKind: raw.EventKind, ResourceID: raw.ResourceID, Minimum: minimum.String()}, nil
	case ProofPossession:
		if scope != ScopeRun || !containsKind(condition, ConditionOwnsGeneratorAtLeast) || !registry.CopyKeys[raw.JustificationCopyKey] || raw.EventKinds != nil || raw.EventKind != "" || raw.ResourceID != "" || raw.Minimum != "" {
			return Proof{}, ErrInvalidCatalog
		}
		return Proof{Kind: raw.Kind, JustificationCopyKey: raw.JustificationCopyKey}, nil
	default:
		return Proof{}, ErrInvalidCatalog
	}
}

func (catalog *Catalog) Definition(id string) (Definition, bool) {
	if catalog == nil {
		return Definition{}, false
	}
	value, ok := catalog.byID[id]
	return cloneDefinition(value), ok
}

func (catalog *Catalog) Score(ids map[string]bool) (int64, error) {
	if catalog == nil || ids == nil {
		return 0, ErrInvalidCatalog
	}
	var total int64
	for id, earned := range ids {
		definition, ok := catalog.byID[id]
		if !earned || !ok || total > decimal.MaxExactInteger-definition.ScoreGrant {
			return 0, ErrInvalidCatalog
		}
		total += definition.ScoreGrant
	}
	return total, nil
}

func containsKind(condition Condition, kind ConditionKind) bool {
	if condition.Kind == kind {
		return true
	}
	for _, child := range condition.Children {
		if containsKind(child, kind) {
			return true
		}
	}
	return false
}

func validRegistry(registry Registry) bool {
	return registry.CopyKeys != nil && registry.GeneratorIDs != nil && registry.EventKinds != nil && registry.ResourceIDs != nil && registry.RunCounters != nil && registry.CareerCounters != nil
}

func decodeStrict(data []byte, target any) error {
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(target); err != nil {
		return err
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		return ErrInvalidCatalog
	}
	return nil
}

func cloneDefinition(source Definition) Definition {
	result := source
	result.Condition = cloneCondition(source.Condition)
	result.Proof.EventKinds = append([]string(nil), source.Proof.EventKinds...)
	return result
}

func cloneCondition(source Condition) Condition {
	result := source
	result.Children = make([]Condition, len(source.Children))
	for index := range source.Children {
		result.Children[index] = cloneCondition(source.Children[index])
	}
	return result
}

func SortedEarnedIDs(ids map[string]bool) []string {
	result := make([]string, 0, len(ids))
	for id, earned := range ids {
		if earned {
			result = append(result, id)
		}
	}
	sort.Strings(result)
	return result
}
